File: Engine/Audio/Platform/Android/Audio.c
/*
 * Copyright (c) 2012-2019 scott.cgi All Rights Reserved.
 *
 * This source code belongs to project Mojoc, which is a pure C Game Engine hosted on GitHub.
 *
 * The Mojoc Game Engine is a serious project with extreme code pursuit,
 * and will continue to iterate,
 * and licensed under the MIT License.
 *
 * License: https://github.com/scottcgi/Mojoc/blob/master/LICENSE
 * GitHub : https://github.com/scottcgi/Mojoc
 *
 * Since  : 2017-2-16
 * Update : 2019-1-7
 * Author : scott.cgi
 */

 
#include "Engine/Toolkit/Platform/Platform.h"


#ifdef IS_PLATFORM_ANDROID


#include <stdlib.h>
#include <SLES/OpenSLES.h>
#include <SLES/OpenSLES_Android.h>

#include "Engine/Toolkit/Utils/ArrayList.h"
#include "Engine/Toolkit/Utils/ArrayStrSet.h"
#include "Engine/Toolkit/Platform/File.h"
#include "Engine/Toolkit/Platform/Log.h"
#include "Engine/Audio/Platform/Audio.h"


// engine interfaces
static SLObjectItf engineObject    = NULL;
static SLEngineItf engineEngine    = NULL;
// output mix interfaces
static SLObjectItf outputMixObject = NULL;


struct AudioPlayer
{
    SLObjectItf object;
    SLPlayItf   play;
    SLSeekItf   seek;
    SLVolumeItf volume;
    const char* filePath;
    int         waitCallbackCount;
};


static ArrayList  (AudioPlayer*) cacheList    [1] = AArrayList_Init (AudioPlayer*, 20);
static ArrayList  (AudioPlayer*) destroyList  [1] = AArrayList_Init (AudioPlayer*, 20);
static ArrayList  (AudioPlayer*) loopList     [1] = AArrayList_Init (AudioPlayer*, 5);
static ArrayList  (AudioPlayer*) testErrorList[1] = AArrayList_Init (AudioPlayer*, 5);
static ArrayStrSet(filePath)     filePathSet  [1] = ArrayStrSet_Init(filePath,     20);


enum
{
    AudioPlayer_WaitMax  = 120,
    AudioPlayer_WaitOver = -1,
};


#define CheckError(tag, filePath)     \
    ALog_A                            \
    (                                 \
        result == SL_RESULT_SUCCESS,  \
        "Audio " tag " failed, "      \
        "error = %x, file path = %s", \
        result,                       \
        filePath                      \
    )


static void Update(float deltaSeconds)
{
    while (destroyList->size > 0)
    {
        AudioPlayer* player = AArrayList_Pop(destroyList, AudioPlayer*);

        // removed from loopList
        AAudio->SetLoop(player, false);

        (*player->object)->Destroy(player->object);

        player->play   = NULL;
        player->volume = NULL;
        player->seek   = NULL;
        player->object = NULL;

        AArrayList_Add(cacheList, player);
    }

    for (int i = testErrorList->size - 1; i > -1 ; --i)
    {
        AudioPlayer* player = AArrayList_Get(testErrorList, i, AudioPlayer*);

        if (player->waitCallbackCount == AudioPlayer_WaitOver)
        {
            AArrayList->RemoveByLast(testErrorList, i);
        }
        else if (player->waitCallbackCount >= AudioPlayer_WaitMax)
        {
            // player callback not called, maybe E/libOpenSLES: Error after prepare: 1
            AArrayList->RemoveByLast(testErrorList, i);
            (*player->object)->Destroy(player->object);
            AArrayList_Add(cacheList, player);

            ALog_E("AAudio player = %s not callback normal", player->filePath);
        }
        else
        {
            ++player->waitCallbackCount;
        }
    }
}


static void SetLoopPause()
{
    for (int i = 0; i < loopList->size; ++i)
    {
        AAudio->Pause(AArrayList_Get(loopList, i, AudioPlayer*));
    }
}


static void SetLoopResume()
{
    for (int i = 0; i < loopList->size; ++i)
    {
        AAudio->Play(AArrayList_Get(loopList, i, AudioPlayer*));
    }
}


static void Init()
{
    SLresult result;

    // create engine
    result = slCreateEngine(&engineObject, 0, NULL, 0, NULL, NULL);
    CheckError("Init slCreateEngine", "");

    // realize the engine
    result = (*engineObject)->Realize(engineObject, SL_BOOLEAN_TRUE);
    CheckError("Init Realize engineObject", "");

    // get the engine interface, which is needed in order to create other objects
    result = (*engineObject)->GetInterface(engineObject, SL_IID_ENGINE, &engineEngine);
    CheckError("Init GetInterface", "");

    result = (*engineEngine)->CreateOutputMix(engineEngine, &outputMixObject, 0, NULL, NULL);
    CheckError("Init CreateOutputMix", "");

    // realize the output mix
    result = (*outputMixObject)->Realize(outputMixObject, SL_BOOLEAN_FALSE);
    CheckError("Init Realize outputMixObject", "");
}


static SLmillisecond inline GetDuration(AudioPlayer* player)
{
    SLresult      result;
    SLmillisecond msec;

    result = (*player->play)->GetDuration(player->play, &msec);
    CheckError("GetDuration", "");

    return msec;
}


static void PlayerCallback(SLPlayItf caller, void* pContext, SLuint32 event)
{
    AudioPlayer* player = (AudioPlayer*) pContext;

    switch (event)
    {
        case SL_PLAYEVENT_HEADATEND:
            // play finish
            AArrayList_Add(destroyList, player);
            (*player->play)->SetPlayState(player->play, SL_PLAYSTATE_PAUSED);
            player->waitCallbackCount = AudioPlayer_WaitOver;
            break;

        case SL_PLAYEVENT_HEADATNEWPOS:
            // play normal
            player->waitCallbackCount = AudioPlayer_WaitOver;
            // remove SL_PLAYEVENT_HEADATNEWPOS for reduce the number of callback
            (*player->play)->SetCallbackEventsMask(player->play, SL_PLAYEVENT_HEADATEND);
            break;

        default:
            break;
    }
}

static inline void InitPlayer(const char* relativeFilePath, AudioPlayer* player)
{
    off_t start;
    off_t length;
    int   fd = AFile->OpenFileDescriptor(relativeFilePath, &start, &length);

    // configure audio source
    SLDataLocator_AndroidFD locFD     = {SL_DATALOCATOR_ANDROIDFD, fd, start, length};
    SLDataFormat_MIME       formatMME = {SL_DATAFORMAT_MIME, NULL, SL_CONTAINERTYPE_UNSPECIFIED};
    SLDataSource            audioSrc  = {&locFD, &formatMME};

    // configure audio sink
    SLDataLocator_OutputMix locOutMix = {SL_DATALOCATOR_OUTPUTMIX, outputMixObject};
    SLDataSink              audioSnk  = {&locOutMix, NULL};

    // create audio player
    SLInterfaceID           ids[3]    = {SL_IID_SEEK,     SL_IID_PLAY,     SL_IID_VOLUME};
    SLboolean               req[3]    = {SL_BOOLEAN_TRUE, SL_BOOLEAN_TRUE, SL_BOOLEAN_TRUE};

    SLresult                result    = (*engineEngine)->CreateAudioPlayer
                                                         (
                                                             engineEngine,
                                                             &player->object,
                                                             &audioSrc,
                                                             &audioSnk,
                                                             3, ids, req
                                                         );

    CheckError("CreatePlayer CreateAudioPlayer", relativeFilePath);

    // realize the player
    result = (*player->object)->Realize(player->object, SL_BOOLEAN_FALSE);
    CheckError("CreatePlayer player Realize", relativeFilePath);

    // get the player interface
    result = (*player->object)->GetInterface(player->object, SL_IID_PLAY, &player->play);
    CheckError("CreatePlayer player GetInterface play", relativeFilePath);

    // player callback
    result = (*player->play)->RegisterCallback(player->play, PlayerCallback, player);
    CheckError("CreatePlayer player RegisterCallback", relativeFilePath);

    result = (*player->play)->SetCallbackEventsMask(player->play,  SL_PLAYEVENT_HEADATEND | SL_PLAYEVENT_HEADATNEWPOS);
    CheckError("CreatePlayer player SetCallbackEventsMask", relativeFilePath);

    // get the seek interface
    result = (*player->object)->GetInterface(player->object, SL_IID_SEEK, &player->seek);
    CheckError("CreatePlayer player GetInterface seek", relativeFilePath);

    // disable looping
    result = (*player->seek)->SetLoop(player->seek, SL_BOOLEAN_FALSE, 0, SL_TIME_UNKNOWN);
    CheckError("CreatePlayer SetLoop", relativeFilePath);

    // get the volume interface
    result = (*player->object)->GetInterface(player->object, SL_IID_VOLUME, &player->volume);
    CheckError("CreatePlayer player GetInterface volume", relativeFilePath);

    player->waitCallbackCount = 0;
    player->filePath          = AArrayStrSet->Get(filePathSet, relativeFilePath);
}


static void SetLoop(AudioPlayer* player, bool isLoop)
{
    SLboolean     isLoopEnabled;
    SLmillisecond startPos;
    SLmillisecond endPos;
    (*player->seek)->GetLoop(player->seek, &isLoopEnabled, &startPos, &endPos);

    if (isLoopEnabled == (SLboolean) isLoop)
    {
        return;
    }

    SLresult result = (*player->seek)->SetLoop(player->seek, (SLboolean) isLoop, 0, GetDuration(player));
    CheckError("SetLoop", player->filePath);

    if (isLoop)
    {
        AArrayList_Add(loopList, player);
    }
    else
    {
        for (int i = 0; i < loopList->size; ++i)
        {
            if (player == AArrayList_Get(loopList, i, AudioPlayer*))
            {
                AArrayList->RemoveByLast(loopList, i);
                break;
            }
        }
    }
}


static void SetVolume(AudioPlayer* player, float volume)
{
    ALog_A
    (
        volume >= 0.0f && volume <= 1.0f,
        "AAudio SetVolume volume %f not in [0.0, 1.0], file path = %s",
        volume,
        player->filePath
    );

    SLresult result = (*player->volume)->SetVolumeLevel(player->volume, (SLmillibel) ((1.0f - volume) * -5000.0f));
    CheckError("SetVolume", player->filePath);
}


static void Play(AudioPlayer* player)
{
    // set the player's state
    SLresult result = (*player->play)->SetPlayState(player->play, SL_PLAYSTATE_PLAYING);
    CheckError("Play", player->filePath);
    AArrayList_Add(testErrorList, player);
}


static void Pause(AudioPlayer* player)
{
    // set the player's state
    SLresult result = (*player->play)->SetPlayState(player->play, SL_PLAYSTATE_PAUSED);
    CheckError("Pause", player->filePath);
}


static void Stop(AudioPlayer* player)
{
    // set the player's state
    SLresult result = (*player->play)->SetPlayState(player->play, SL_PLAYSTATE_STOPPED);
    CheckError("Stop", player->filePath);
    AArrayList_Add(destroyList, player);
    player->waitCallbackCount = AudioPlayer_WaitOver;
}


static bool IsPlaying(AudioPlayer* player)
{
    SLuint32 state;
    (*player->play)->GetPlayState(player->play, &state);

    if (state == SL_PLAYSTATE_PLAYING)
    {
        return true;
    }
    else
    {
        return false;
    }
}


static AudioPlayer* GetPlayer(const char* relativeFilePath)
{
    AudioPlayer* player = AArrayList_Pop(cacheList, AudioPlayer*);

    if (player == NULL)
    {
        player = malloc(sizeof(AudioPlayer));
    }

    InitPlayer(relativeFilePath, player);

    return player;
}


static void Release()
{
    (*engineObject)   ->Destroy(engineObject);
    (*outputMixObject)->Destroy(outputMixObject);

    engineObject    = NULL;
    engineEngine    = NULL;
    outputMixObject = NULL;

    AArrayList  ->Release(cacheList);
    AArrayList  ->Release(destroyList);
    AArrayList  ->Release(loopList);
    AArrayList  ->Release(testErrorList);
    AArrayStrSet->Release(filePathSet);
}

struct AAudio AAudio[1] =
{
    Init,
    Release,
    Update,
    SetLoopPause,
    SetLoopResume,
    GetPlayer,

    SetVolume,
    SetLoop,

    Play,
    Pause,
    Stop,
    IsPlaying,
};


#endif

File: Engine/Audio/Platform/IOS/Audio.m
/*
 * Copyright (c) 2012-2019 scott.cgi All Rights Reserved.
 *
 * This source code belongs to project Mojoc, which is a pure C Game Engine hosted on GitHub.
 * The Mojoc Game Engine is licensed under the MIT License, and will continue to be iterated with coding passion.
 *
 * License  : https://github.com/scottcgi/Mojoc/blob/master/LICENSE
 * GitHub   : https://github.com/scottcgi/Mojoc
 * CodeStyle: https://github.com/scottcgi/Mojoc/wiki/Code-Style
 *
 * Since    : 2017-5-4
 * Update   : 2019-1-7
 * Author   : scott.cgi
 */


#include "Engine/Toolkit/Platform/Platform.h"


//--------------------
#ifdef IS_PLATFORM_IOS
//--------------------


#include <stddef.h>
#include <Foundation/Foundation.h>
#include <AudioToolbox/AudioToolbox.h>
#include <OpenAL/OpenAL.h>

#include "Engine/Audio/Platform/Audio.h"
#include "Engine/Toolkit/Platform/Log.h"
#include "Engine/Toolkit/Utils/ArrayStrMap.h"
#include "Engine/Toolkit/Utils/ArrayStrSet.h"


typedef struct
{
    ALsizei size;
    ALenum  format;
    ALsizei sampleRate;
    void*   data;
}
AudioData;


static ArrayStrMap(filePath, AudioData*) audioDataMap[1] = AArrayStrMap_Init(AudioData*,  20);
static ArrayList  (AudioPlayer*)         cacheList  [1] = AArrayList_Init  (AudioPlayer*, 20);
static ArrayList  (AudioPlayer*)         destroyList[1] = AArrayList_Init  (AudioPlayer*, 20);
static ArrayList  (AudioPlayer*)         loopList   [1] = AArrayList_Init  (AudioPlayer*, 5 );
static ArrayStrSet(filePath)             filePathSet[1] = ArrayStrSet_Init (filePath,     20);


#define CheckAudioDataError(tag)               \
    ALog_A                                     \
    (                                          \
        error == noErr,                        \
        "AAudio GetAudioData " tag " failed, " \
        "OSStatus error = %x, file path = %s", \
        (int) error,                           \
        relativeFilePath                       \
    )

#define CheckAudioError(tag, filePath)        \
    ALog_A                                    \
    (                                         \
        (error = alGetError()) == AL_NO_ERROR,\
        "AAudio " tag " failed, "             \
        "alGetError = %x, file path = %s",    \
        error,                                \
        filePath                              \
    )


/**
 * The return data will cached in AudioDataMap.
 */
static inline AudioData* GetAudioData(const char* relativeFilePath)
{
    AudioData* audioData = NULL;
    int        index     = AArrayStrMap->GetIndex(audioDataMap, relativeFilePath);

    if (index < 0)
    {
        AudioStreamBasicDescription fileFormat;
        AudioStreamBasicDescription outputFormat;

        SInt64          fileLengthInFrames = 0;
        UInt32          propertySize       = sizeof(fileFormat);
        ExtAudioFileRef audioFileRef       = NULL;
        NSString*       path               = [[NSBundle mainBundle] pathForResource:
                                             [NSString stringWithUTF8String:relativeFilePath] ofType:nil];

        CFURLRef        fileUrl            = CFURLCreateWithString(kCFAllocatorDefault, (CFStringRef) path, NULL);
        OSStatus        error              = ExtAudioFileOpenURL(fileUrl, &audioFileRef);
        CheckAudioDataError("ExtAudioFileOpenURL");

        CFRelease(fileUrl);

        // get the audio data format
        error = ExtAudioFileGetProperty(audioFileRef, kExtAudioFileProperty_FileDataFormat, &propertySize, &fileFormat);
        CheckAudioDataError("ExtAudioFileGetProperty kExtAudioFileProperty_FileDataFormat");

        if (fileFormat.mChannelsPerFrame > 2)
        {
            ALog_E
            (
                "AAudio GetAudioData unsupported format,"
                "channel count = %u is greater than stereo, relativeFilePath = %s",
                (unsigned int) fileFormat.mChannelsPerFrame,
                relativeFilePath
            );
        }

        // set the client format to 16 bit signed integer (native-endian) data
        // maintain the channel count and sample rate of the original source format
        outputFormat.mSampleRate       = fileFormat.mSampleRate;
        outputFormat.mChannelsPerFrame = fileFormat.mChannelsPerFrame;
        outputFormat.mFormatID         = kAudioFormatLinearPCM;
        outputFormat.mBytesPerPacket   = outputFormat.mChannelsPerFrame * 2;
        outputFormat.mFramesPerPacket  = 1;
        outputFormat.mBytesPerFrame    = outputFormat.mChannelsPerFrame * 2;
        outputFormat.mBitsPerChannel   = 16;
        outputFormat.mFormatFlags      = kAudioFormatFlagsNativeEndian  |
                                         kAudioFormatFlagIsPacked       |
                                         kAudioFormatFlagIsSignedInteger;

        // set the desired client (output) data format
        error = ExtAudioFileSetProperty
                (
                    audioFileRef,
                    kExtAudioFileProperty_ClientDataFormat,
                    sizeof(outputFormat),
                    &outputFormat
                );
        CheckAudioDataError("ExtAudioFileSetProperty kExtAudioFileProperty_ClientDataFormat");

        // get the total frame count
        propertySize = sizeof(fileLengthInFrames);
        error        = ExtAudioFileGetProperty
                       (
                           audioFileRef,
                           kExtAudioFileProperty_FileLengthFrames,
                           &propertySize,
                           &fileLengthInFrames
                       );
        CheckAudioDataError("ExtAudioFileGetProperty kExtAudioFileProperty_FileLengthFrames");

        // read all the data into memory
        UInt32 framesToRead = (UInt32) fileLengthInFrames;
        UInt32 dataSize     = framesToRead * outputFormat.mBytesPerFrame;

//----------------------------------------------------------------------------------------------------------------------

        audioData             = malloc(sizeof(AudioData) + dataSize);
        audioData->size       = (ALsizei) dataSize;
        audioData->format     = outputFormat.mChannelsPerFrame > 1 ? AL_FORMAT_STEREO16 : AL_FORMAT_MONO16;
        audioData->sampleRate = (ALsizei) outputFormat.mSampleRate;
        audioData->data       = (char*) audioData + sizeof(AudioData);

        AudioBufferList dataBuffer;
        dataBuffer.mNumberBuffers              = 1;
        dataBuffer.mBuffers[0].mDataByteSize   = dataSize;
        dataBuffer.mBuffers[0].mNumberChannels = outputFormat.mChannelsPerFrame;
        dataBuffer.mBuffers[0].mData           = audioData->data;

        // read the data into an AudioBufferList
        error = ExtAudioFileRead(audioFileRef, &framesToRead, &dataBuffer);
        CheckAudioDataError("ExtAudioFileRead");

        // dispose the ExtAudioFileRef, it is no longer needed
        ExtAudioFileDispose(audioFileRef);

        AArrayStrMap_InsertAt(audioDataMap, relativeFilePath, -index - 1, audioData);
    }
    else
    {
        audioData = AArrayStrMap_GetAt(audioDataMap, index, AudioData*);
    }

    return audioData;
}


#undef CheckAudioDataError


//----------------------------------------------------------------------------------------------------------------------


static ALCdevice*                device                 = NULL;
static ALCcontext*               context                = NULL;
static alBufferDataStaticProcPtr alBufferDataStaticProc = NULL;


/**
 * Defined in Audio.h
 */
struct AudioPlayer
{
    ALuint      sourceId;
    ALuint      bufferId;
    const char* filePath;
};


static void Update(float deltaSeconds)
{
    for (int i = destroyList->size - 1; i > -1; --i)
    {
        AudioPlayer* player = AArrayList_Get(destroyList, i, AudioPlayer*);
        
        ALint state;
        alGetSourcei(player->sourceId, AL_SOURCE_STATE, &state);
        
        if (state == AL_STOPPED)
        {
            alDeleteSources(1, &player->sourceId);
            alDeleteBuffers(1, &player->bufferId);

            player->sourceId = 0;
            player->bufferId = 0;
            player->filePath = NULL;

            AArrayList->Remove(destroyList, i);
            AArrayList_Add(cacheList, player);
        }
    }
}


static void SetLoopPause()
{
    for (int i = 0; i < loopList->size; ++i)
    {
        AAudio->Pause(AArrayList_Get(loopList, i, AudioPlayer*));
    }
}


static void SetLoopResume()
{
    for (int i = 0; i < loopList->size; ++i)
    {
        AAudio->Play(AArrayList_Get(loopList, i, AudioPlayer*));
    }
}


static void Init()
{
    // get static buffer data API
    alBufferDataStaticProc = alcGetProcAddress(NULL, "alBufferDataStatic");
    
    // create a new OpenAL Device
    // pass NULL to specify the system’s default output device
    device = alcOpenDevice(NULL);
    
    if (device != NULL)
    {
        // create a new OpenAL Context
        // the new context will render to the OpenAL Device just created
        context = alcCreateContext(device, 0);
        
        if (context != NULL)
        {
            // make the new context the Current OpenAL Context
            alcMakeContextCurrent(context);
        }
    }
    else
    {
        ALog_E("AAudio Init failed, OpenAL cannot open device");
    }
    
    // clear any errors
    ALenum error;
    CheckAudioError("Init", "");
}


static inline void InitPlayer(const char* relativeFilePath, AudioPlayer* player)
{
    ALenum     error;
    AudioData* audioData = GetAudioData(relativeFilePath);
    
    alGenBuffers(1, &player->bufferId);
    CheckAudioError("InitPlayer generate buffer", relativeFilePath);
    
    // use the static buffer data API
    // the data will not copy in buffer so cannot free data until buffer deleted
    alBufferDataStaticProc(player->bufferId, audioData->format, audioData->data, audioData->size, audioData->sampleRate);
    CheckAudioError("InitPlayer attach audio data to buffer", relativeFilePath);

    alGenSources(1, &player->sourceId);
    CheckAudioError("InitPlayer generate source", relativeFilePath);

    // turn Looping off
    alSourcei(player->sourceId,  AL_LOOPING, AL_FALSE);
    
    // set Source Position
    alSourcefv(player->sourceId, AL_POSITION, (const ALfloat[3]) {0.0f, 0.0f, 0.0f});
    
    // set source reference distance
    alSourcef(player->sourceId,  AL_REFERENCE_DISTANCE, 0.0f);
    
    // attach OpenAL buffer to OpenAL Source
    alSourcei(player->sourceId,  AL_BUFFER, player->bufferId);
    CheckAudioError("InitPlayer attach buffer to source", relativeFilePath);

    // set player name
    player->filePath = AArrayStrSet->Get(filePathSet, relativeFilePath);
}


static void SetLoop(AudioPlayer* player, bool isLoop)
{
    ALint isLoopEnabled;
    alGetSourcei(player->sourceId, AL_LOOPING, &isLoopEnabled);
    
    if (isLoopEnabled == isLoop)
    {
        return;
    }
    
    alSourcei(player->sourceId, AL_LOOPING, (ALint) isLoop);

    // move player from removeList to addList
    
    ArrayList(AudioPlayer*)* addList;
    ArrayList(AudioPlayer*)* removeList;

    if (isLoop)
    {
        addList    = loopList;
        removeList = destroyList;
    }
    else
    {
        addList    = destroyList;
        removeList = loopList;
    }

    for (int i = 0; i < removeList->size; ++i)
    {
        if (player == AArrayList_Get(removeList, i, AudioPlayer*))
        {
            AArrayList->RemoveByLast(removeList, i);
            break;
        }
    }

    AArrayList_Add(addList, player);
}


static void SetVolume(AudioPlayer* player, float volume)
{
    ALog_A
    (
        volume >= 0.0f && volume <= 1.0f,
        "AAudio SetVolume volume %f not in [0.0, 1.0], audio file path = %s",
        volume,
        player->filePath
    );

    alSourcef(player->sourceId, AL_GAIN, volume);
    
    ALenum error;
    CheckAudioError("SetVolume", player->filePath);
}


static void Play(AudioPlayer* player)
{
    alSourcePlay(player->sourceId);
    ALenum error;
    CheckAudioError("Play", player->filePath);
}


static void Pause(AudioPlayer* player)
{
    alSourcePause(player->sourceId);
    ALenum error;
    CheckAudioError("Pause", player->filePath);
}


static void Stop(AudioPlayer* player)
{
    // after alSourceStop call alDeleteSources will cause OpenAL leak OALSource:AddPlaybackMessage
    alSourceStop(player->sourceId);
    ALenum error;
    CheckAudioError("Stop", player->filePath);

    // put player into destroyList
    SetLoop(player, false);
}

static bool IsPlaying(AudioPlayer* player)
{
    ALint state;
    alGetSourcei(player->sourceId, AL_SOURCE_STATE, &state);
    return state == AL_PLAYING;
}


static AudioPlayer* GetPlayer(const char* relativeFilePath)
{
    AudioPlayer* player = AArrayList_Pop(cacheList, AudioPlayer*);
    
    if (player == NULL)
    {
        player = malloc(sizeof(AudioPlayer));
    }
    
    InitPlayer(relativeFilePath, player);
    AArrayList_Add(destroyList, player);
    
    return player;
}


#undef CheckAudioError


static void Release()
{
    // release context
    alcDestroyContext(context);
    // close device
    alcCloseDevice(device);

    AArrayStrMap->Release(audioDataMap);
    AArrayList  ->Release(cacheList);
    AArrayList  ->Release(destroyList);
    AArrayList  ->Release(loopList);
    AArrayStrSet->Release(filePathSet);
}


struct AAudio AAudio[1] =
{{
    Init,
    Release,
    Update,
    SetLoopPause,
    SetLoopResume,
    GetPlayer,

    SetVolume,
    SetLoop,

    Play,
    Pause,
    Stop,
    IsPlaying,
}};


//-----------------------
#endif // IS_PLATFORM_IOS
//-----------------------
